test/utils.jl

```diff
@@ -340,18 +340,136 @@ end
+    m2 = Chain(Conv((3, 3), 3 => 16), BatchNorm(16), x -> reshape(x, :, size(x)[end]), Dropout(0.1))
+    m2[2].μ .= rand(Float32, size(m2[2].μ)...)
+    loadmodel!(m1, m2)
+    # non-trainable parameters are copied as well
+    @test m1[2].μ == m2[2].μ
+    # functions are not copied
+    @test m1[3] == Flux.flatten
+    # dropout rate is not copied
+    @test m1[4].p == 0.2
+
+    # from LegolasFlux (https://github.com/beacon-biosignals/LegolasFlux.jl/blob/80569ab63a8248a8a063c76e0bbf701f4ada9bd4/examples/digits.jl#L33)
+    # tests Chain(...) vs Chain([...])
+    # tests MaxPool
+    # tests testmode!/trainmode! is not copied
+    # tests Dense, Conv, BatchNorm, Dropout (like above) but in a bigger model
+    chain1 = Chain(Dropout(0.2),
+                   Conv((3, 3), 1 => 32, relu),
+                   BatchNorm(32, relu),
+                   MaxPool((2, 2)),
+                   Dropout(0.2),
+                   Conv((3, 3), 32 => 16, relu),
+                   Dropout(0.2),
+                   MaxPool((2, 2)),
+                   Dropout(0.2),
+                   Conv((3, 3), 16 => 10, relu),
+                   Dropout(0.2),
+                   x -> reshape(x, :, size(x, 4)),
+                   Dropout(0.2),
+                   Dense(90, 10),
+                   softmax)
+    chain2 = Chain([Dropout(0.1),
+                   Conv((3, 3), 1 => 32, relu),
+                   BatchNorm(32, relu),
+                   MaxPool((3, 3)),
+                   Dropout(0.1),
+                   Conv((3, 3), 32 => 16, relu),
+                   Dropout(0.1),
+                   MaxPool((3, 3)),
+                   Dropout(0.1),
+                   Conv((3, 3), 16 => 10, relu),
+                   Dropout(0.1),
+                   x -> reshape(x, :, size(x, 4)),
+                   Dropout(0.1),
+                   Dense(90, 10),
+                   softmax])
+    chain2[3].μ .= 5f0
+    chain2[3].σ² .= 2f0
+    testmode!(chain2)
+    loadmodel!(chain1, chain2)
+    for (dst, src) in zip(chain1, chain2)
+      if dst isa Dropout
+        @test dst.p == 0.2
+      elseif dst isa Union{Conv, Dense}
+        @test dst.weight == src.weight
+        @test dst.bias == src.bias
+      elseif dst isa MaxPool
+        @test dst.k == (2, 2)
+      elseif dst isa BatchNorm
+        @test dst.μ == src.μ
+        @test dst.σ² == src.σ²
+        @test isnothing(dst.active)
+      end
+    end
+
+    # copy only a subset of the model
+    chain1[end - 1].weight .= 1f0
+    chain1[3].μ .= 3f0
+    chain1[2].bias .= 5f0
+    loadmodel!(chain2[end - 1], chain1[end - 1])
+    loadmodel!(chain2[3], chain1[3])
+    @test chain2[end - 1].weight == chain1[end - 1].weight
+    @test chain2[3].μ == chain1[3].μ
+    @test chain2[2].bias != chain1[2].bias
+
+    # test shared weights
+    m1 = Chain(Dense(10 => 5), Dense(5 => 2))
+    m2 = Chain(Dense(transpose(m1[2].weight)), Dense(permutedims(m1[1].weight)))
+    m3 = Chain(Dense(m1[1].weight), Dense(m1[2].weight))
```

Comment:
Can you say what you mean by shared here? I don't think the fmap cache would see any duplicates here.

Reply:
Okay I have a better idea of what you meant by shared weights now. Something like the model below, but the tests I wrote just document the current behavior. What do we want? If someone loads an untied set of weights into a set of tied weights, should we warn or error?

> fmap cache

Note that there is no longer `fmap` here, so it's caching behaviors are unused. This kind of diversion is how we recurse things is why I prefer to use `fmap` over manually using `functor`. But there is no way to hold "one level up" in the error message with `fmap` (at least not that I could figure out).

```suggestion
    # test shared weights
    encoder = Chain(Dense(10 => 5), Dense(5 => 2))
    decoder = Chain(Dense(transpose(m1[2].weight)), Dense(permutedims(m1[1].weight)))
    m1 = Chain(encoder, decoder)
    m2 = Chain(Chain(Dense(10 => 5), Dense(5 => 2)), Chain(Dense(2 => 5), Dense(5 => 10)))
    loadmodel!(m1, m2)
    # what behavior do we want
    @test encoder[2].weight === parent(decoder[1].weight)
    @test encoder[2].weight == permutedims(m2[2][1].weight)
    @test encoder[1].weight == m2[1].weight
    @test encoder[1].weight != permutedims(decoder[2].weight)
```

Reply:
Erroring seems like the correct thing to me too.

Reply:
My quick attempt to write this, starting from `fmap` and adapting...
```julia
function modelcopy!(x, y; cache = IdDict())
  if haskey(cache, x)
    # If we've seen (and thus updated) x already, check that new y is the same, or nothing:
    leaf_tiecheck(x, y)
    return cache[x]
  end
  cache[x] = if Functors.isleaf(x)
    leafcopy!(x, y)
  else
    # _default_walk has mismarch errors from functor(T, y), not as pretty as could be
    Functors._default_walk((x, y) -> modelcopy!(x, y; cache), x, y)
  end
end

# Main case, error on immutable array x (at present):
leafcopy!(x::AbstractArray, y::AbstractArray) = x .= y  # demands equal size, could make a prettier error
leaf_tiecheck(x::AbstractArray, y::AbstractArray) = x == y || error("tied weights in `dst` must be identical in `src`, or else `nothing`")

# Bool represents absent bias, equivalent to zero bias:
leafcopy!(x::AbstractArray, y::Bool) = x .= y
leafcopy!(x::Bool, y::AbstractArray) = iszero(y) ? x : error("can't copy nonzero bias in `src` to absent bias in `dst`")
leaf_tiecheck(x::AbstractArray, y::Bool) = iszero(x) || error("tied weights in `dst` cannot have some zero/absent bias, others nonzero")
leaf_tiecheck(x::Bool, y::AbstractArray) = iszero(x) || error("tied weights in `dst` cannot have some zero/absent bias, others nonzero")

# Fallback, for e.g. activation functions, but also missing nodes:
leafcopy!(x, y) = x
leaf_tiecheck(x, y) = nothing


# This fmapstructure produces a tree of tuples, with all ties removed, ideal for saving?
sh = [1,2]; m = (x=sh, y=[3,4], z=sh, b=false, a=sin);
mflat = fmapstructure(identity, m; prune=nothing)
# That is accepted as src by modelcopy!
modelcopy!(m, (x = [1, 20], y = [300, 4000], z = nothing, b=[0,0], a=nothing))
modelcopy!(m, (x=[10,2], y=[3,4], z=[10,2], b=false, a=nothing))  # also ok
modelcopy!(m, (x=nothing, y=[30,40], z=nothing, b=nothing, a=nothing))  # doesn't change m.x
# Mismatch of a weight which should be shared is an error:
modelcopy!(m, (x=[1,2], y=[3,4], z=[100,200], b=nothing, a=nothing))

# Could simplify further:
function simpletree(m)
    fmapstructure(m; prune=nothing) do x
        # We know isleaf(x), but further keep only values modelcopy! will accept:
        x isa AbstractArray && return x
        x === false && return x
        nothing
    end
end
simpletree(m)  # has a = nothing
```